codegenerator/cli/templates/static/codegen/src/LoadLayer.res

```diff
@@ -22,42 +26,48 @@ module LoadActionMap = {
     fieldValue: Utils.magic(fieldValue),
   }
 
-  type loadIndex<'entity> = {
+  type loadIndex = {
     index: TableIndices.Index.t,
-    loadArgs: loadArgs<'entity>,
-    loadCallbacks: array<loadMultiple<'entity>>,
+    loadArgs,
+    loadCallbacks: array<loadMultiple>,
   }
 
-  type key = string
   type t = {
-    singleEntities: dict<array<loadSingle<Entities.internalEntity>>>,
-    lookupByIndex: dict<loadIndex<Entities.internalEntity>>,
+    byId: dict<loadSingle>,
+    lookupByIndex: dict<loadIndex>,
     entityMod: module(Entities.InternalEntity),
     logger: Pino.t,
   }
   let make = (~entityMod, ~logger) => {
-    singleEntities: Js.Dict.empty(),
+    byId: Js.Dict.empty(),
     lookupByIndex: Js.Dict.empty(),
     entityMod,
     logger,
   }
-  let getSingleEntityIds = (map: t) => map.singleEntities->Js.Dict.keys
+  let getIdsToLoad = (map: t) => map.byId->Js.Dict.keys
   let getIndexes = (map: t) => map.lookupByIndex->Js.Dict.values
-  let entriesSingleEntities: t => array<(key, array<loadSingle<'entity>>)> = map =>
-    map.singleEntities->Js.Dict.entries
 
-  let addSingle = (map: t, ~entityId, ~resolve) => {
-    let loadCallback: loadSingle<'entity> = {
-      resolve: resolve,
+  let registerLoadById = (map: t, ~entityId) => {
+    switch map.byId->Utils.Dict.dangerouslyGetNonOption(entityId) {
+    | Some(loadRecord) => loadRecord.promise
+    | None => {
+      let promise = Promise.make((resolve, reject) => {
+        let loadRecord: loadSingle = {
+          resolve,
+          reject,
+          promise: %raw(`null`)
+        }
+        map.byId->Js.Dict.set(entityId, loadRecord)
+      })
+      // Don't use ref, since it'll allocate an object to store .contents
+      (map.byId->Js.Dict.unsafeGet(entityId)).promise = promise
```

Comment:
I made something very similar in utils FYI. (Called promiseWithHandles I think)

Reply:
I'll take a look

Reply:
I've checked it out, but decided to remove `promiseWithHandles` instead. I found it less performant, and since it's not used anywhere, I thought that it's better not to have the abstraction.